Stories: Fix Media Import bugs

bjtitus opened the pull request:
* Fixes a bug where media was unnecessarily exported (instead of copied/reused) when adding to story.
* Adds a Loading indicator during media import so there is some indicator while media files are being exported.
* Allows video uploads for local URLs so that Stories media can be uploaded to the media library.

### Testing
* Go to the Blog Details page
* Create a new Story Post
* Add several pieces of media (with at least 1 video) from your device
* Ensure that media is imported quickly into the Stories editor
* Publish the post
* Ensure that video is uploaded and published in story

## Code Overview



PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

bjtitus committed:
Fix media exporting file move + add loading

Copies the exported media file instead of moving and shows a loading indicator in case the media needs to export.

bjtitus committed:
Allow video files to be uploaded on all sites

This was only allowed for PHAsset before but we want to allow for any media.
Android doesn't even use this property so we should be safe.

bjtitus commented on WordPress/Classes/Utility/Media/MediaURLExporter.swift:
```diff
@@ -77,10 +77,9 @@ class MediaURLExporter: MediaExporter {
-            if !urlExportOptions.allowableFileExtensions.contains(fileExtension) {
+        if let fileExtension = fileURL.typeIdentifierFileExtension {
+            // Check the default file types. We want to limit to supported types but mobile is not restricted to only allowed ones.
+            if !MediaImportService.defaultAllowableFileExtensions.contains(fileExtension) {
```
Checked with @aforcier that Android does not appear to be using the `allowed_file_types` from the blog.

On iOS, we don't check the list for PHAssets but do for media which comes from a URL. This needs to change for Stories since we are uploading video which came from Photos (or WP Media) and then ended up re-exported with modifications.

The `MediaImportService.defaultAllowableFileExtensions` value is still used so that we still check for supported file types.

bjtitus commented on WordPress/Classes/Services/Stories/WPMediaPicker+MediaPicker.swift:
```diff
@@ -161,38 +165,52 @@ extension WPMediaAsset {
                     return URLSession.shared.downloadTaskPublisher(url: assetURL).tryMap { (location, _) -> URL in
                         if let location = location {
-                            try FileManager.default.moveItem(at: location, to: url)
+                            try FileManager.default.moveItem(at: location, to: exportURL)
```
It is safe to move the item here since we just downloaded it.

bjtitus commented on WordPress/Classes/Services/Stories/WPMediaPicker+MediaPicker.swift:
```diff
@@ -161,38 +165,52 @@ extension WPMediaAsset {
-                try FileManager.default.moveItem(at: assetURL, to: url)
-                return Just(url).setFailureType(to: Error.self).eraseToAnyPublisher()
+                do {
+                    try FileManager.default.copyItem(at: assetURL, to: exportURL)
```
We must copy the item instead of moving because we don't have permission to move files from the Photos library.

bjtitus commented on WordPress/Classes/Services/Stories/WPMediaPicker+MediaPicker.swift:
```diff
@@ -161,38 +165,52 @@ extension WPMediaAsset {
 
             if let assetURL = urlAsset?.url {
+                let existingExtension = assetURL.pathExtension
+                let exportURL = url.appendingPathExtension(existingExtension)
```
Uses the Media's original extension instead of hardcoding it.

aerych replied:
Maybe `url.appendingPathExtension(assetURL.pathExtension)` since we don't use `existingExtension` for anything else?  Up to you.

bjtitus committed:
Use assetURL  + export to Media Cache directory

aerych commented on WordPress/Classes/Services/Stories/WPMediaPicker+MediaPicker.swift:
```diff
@@ -161,38 +165,50 @@ extension WPMediaAsset {
         return videoAssetPublisher().tryMap { asset -> AnyPublisher<URL, Error> in
             let filename = UUID().uuidString
-            let url = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true).appendingPathComponent(filename).appendingPathExtension("mp4")
+            let storiesURL = try MediaFileManager.cache.directoryURL().appendingPathComponent("Stories", isDirectory: true)
```
It doesn't look like "Stories" is used anywhere else but are there plans to?  Maybe make this a constant if so?

bjtitus replied:
I decided to reuse this directory name for the `saveDirectory` used in `StoryEditor` in b29af92. This should keep the Stories files separated out in case we need to do any cleanup or rearranging later on.

`saveDirectory` is technically temporary (storing archives before media is uploaded and receives its remote media ID), but we want to be in control so we aren't using the system temp directory.

aerych reviewed (commented):
Heya @bjtitus 👋 

I had one or two nitpicks/questions code wise. I need to brush up on my Combine. :) 

I tested on an iPhone 11 Pro device running 14.4.  There were a few things I noticed that I wanted to share. 

- During one test, I selected three images and two videos. The HUD stayed up for a long time
and I noticed that the preview button was tappable... so I tapped it.  I was then stuck on a different screen with no way back.  I didn't think to get a screenshot and I haven't been able to reproduce.  Just FYI. 
- I noticed that if I'm on the blog details screen when I start a story, when I go to publish the story, it appears to fail silently.  It looks like media is uploaded but not the post. The post never appears in the post list and I don't see the post publish nudges.  However, when I create a story from the Post list, everything works.  I can also create a regular post from the blog detail screen just fine. 
- Video frames load super fast.  Image frames are very slow by comparison. 

Thoughts on these?

bjtitus committed:
Add media resizing upon import

Stories only needs images roughly the size of the screen. We may update with higher resolutions later but this keeps performance in line and reduces export and upload times.

bjtitus committed:
Hide Picker's inputAccessoryView while exporting

Also adds an error message if the export fails and restores the selected media items.

bjtitus committed:
Addresses PR suggestions

bjtitus committed:
Reconfigure Stories media URLs

bjtitus committed:
Fix saveDirectory usage

bjtitus committed:
Remove media file URL unwrap + add error

Also adds docs for directory properties

bjtitus commented:
Thanks @aerych!

> The HUD stayed up for a long time and I noticed that the preview button was tappable

I've hidden those buttons while the HUD is up and added some error handling if the import does fail for some reason. It may be hard to get it to happen again since the image resizing change (mentioned below) should make import much quicker.

> I noticed that if I'm on the blog details screen when I start a story, when I go to publish the story, it appears to fail silently.

🤔 I'm not sure why this would happen and I'm having trouble reproducing.

> Image frames are very slow by comparison

I've made a modification to resize images to match the screen size. This mirror Android's behavior and the behavior of the camera UI. Should be much quicker now! This changes includes a duplicate of the `image(with size:...` method in `WPMediaPicker`, which I will migrate over in a new PR tomorrow. The resized image doesn't always match the dimensions exactly so we need to wait for the non-degraded version

aerych commented on WordPress/Classes/ViewRelated/Gutenberg/Processors/GutenbergMediaFilesUploadProcessor.swift:
```diff
@@ -35,7 +35,11 @@ class GutenbergMediaFilesUploadProcessor: Processor {
+                if let mediaURL = URL(string: mediaFile.url) {
+                    try FileManager.default.moveItem(at: mediaURL, to: newURL)
+                } else {
+                    DDLogError("No Media File URL was present. This is Stories error and should be investigated.")
```
`This is Stories error and should be investigated` > `This is a Stories error and should be investigated`

aerych reviewed (approved):
Heya @bjtitus 👋 
Changes look good from here.  Image frames are much snappier and seem consistent with video loading.  I wasn't able to reproduce the issue with launching the stories editor from the blog details screen.  Not sure what that was about. 
I had one nitpick about logging but otherwise :shipit: from me after resolving the merge conflict. :)

bjtitus committed:
PR feedback and typo fixes

bjtitus committed:
Merge branch 'release/16.8' into stories/media-import

# Conflicts:
#	WordPress/Classes/Services/Stories/StoryEditor.swift
